gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java

```diff
@@ -257,6 +257,20 @@ public T construct() {
     };
   }
 
+  private static boolean hasStringKeyType(Type mapType) {
+    // If mapType is not parameterized, cannot assume that key is String, because if it
+    // is not String and does not implement Comparable it would lead to a ClassCastException
+    if (!(mapType instanceof ParameterizedType)) {
+      return false;
+    }
```

Comment:
This would prevent `LinkedTreeMap` from being used for raw map types, which should be invisible to the user (since `LnkedTreeMap` is an internal class) but would disable the DoS protection in that case for Java 7. Not sure if that would be acceptable, but on the other hand it prevents potential `ClassCastException`s.

Reply:
This is kind of a big change, isn't it? It seems to me that `gson.fromJson(s, Map.class)` is a pretty common idiom and we're changing what it does. Agreed that people shouldn't be relying on getting a `LinkedTreeMap`, but is it something they could observe indirectly? For example because of memory usage. Though it looks to me as if `LinkedTreeMap` is fairly expensive in memory terms because of all those fields in its `Node` class.

I see some other options. I think the whole situation is a bit of a mess to be honest, but here are some ways we *might* be able to clean it up a bit:

1. Keep the approach here: raw `Map` never deserializes into a `LinkedTreeMap`, but some parameterized `Type` instances will. (We can work on adjusting exactly which ones.)
2. Never deserialize into `LinkedTreeMap` here. We would still use the class for `JsonObject` but here we would always use `LinkedHashMap`.
3. Give `LinkedTreeMap` a fallback mode, where if a key is added that is not `Comparable`, or that can't be compared to existing keys, it copies the existing entries into a private `LinkedHashMap` and uses that for all subsequent operations.

I suppose my underlying question is why `LinkedTreeMap` exists in the first place. What problem does it solve? It seems to me that it is usually going to be both bigger and less performant than `LinkedHashMap`, though I haven't measured that.

Reply:
Below where this method `hasStringKeyType` is called, I switched the `if` statement branches, which is why it might be a bit confusing.

I think the intended logic here is:
If it is a `Map<String, ...>` use Gson's `LinkedTreeMap` to protect against potential DoS for JDK's `LinkedHashMap` in Java < 8 ([JDK-8046170](https://bugs.openjdk.org/browse/JDK-8046170)). Otherwise use JDK's `LinkedHashMap`.

The check below was previously (roughly):
```java
if (type instanceof ParameterizedType && !(String.class.isAssignableFrom(
          TypeToken.get(((ParameterizedType) type).getActualTypeArguments()[0]).getRawType()))) {
    ...
} else {
    return new LinkedTreeMap<>();
}
```
For raw `Map.class` and for any parameterized type `Map<String, ...>` it would have used `LinkedTreeMap`.
However, using `LinkedTreeMap` for raw `Map.class` can lead to `ClassCastException`s, as shown by the examples in #1247.

Therefore the changes by this PR actually reduce the cases where `LinkedTreeMap` is used (which was the concern I wanted to express with my original comment here).
But if you think that is fine, then I guess we can leave it like this.